flang/lib/Lower/NumericRuntime.cpp

```diff
@@ -0,0 +1,102 @@
+//===-- NumericRuntime.cpp -- runtime for numeric intrinsics -------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "../../runtime/numeric.h"
+#include "RTBuilder.h"
+#include "flang/Lower/Bridge.h"
+#include "flang/Lower/CharacterExpr.h"
+#include "flang/Lower/FIRBuilder.h"
+#include "flang/Lower/Support/BoxValue.h"
+#include "flang/Lower/Todo.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "flang/Lower/NumericRuntime.h"
+
+using namespace Fortran::runtime;
+
+/// Helper function to recover the KIND from the REAL FIR type.
+static int discoverRealKind(mlir::Type fltTy) {
+  if (fltTy.isF16())
+    return 2;
+  if (fltTy.isBF16())
+    return 3;
+  if (fltTy.isF32())
+    return 4;
+  if (fltTy.isF64())
+    return 8;
+  if (fltTy.isF80())
+    return 10;
+  if (fltTy.isF128())
+    return 16;
+  llvm_unreachable("unexpected type");
+}
+
+/// Generate call to RRSpacing intrinsic runtime routine. 
+mlir::Value
+Fortran::lower::genRRSpacing(Fortran::lower::FirOpBuilder &builder, 
+                           mlir::Location loc, mlir::Value x) { 
+  mlir::FuncOp func;
+  switch (discoverRealKind(x.getType())) {
+  case 4:
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(RRSpacing4)>(loc, builder);
+    break;
+  case 8:
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(RRSpacing8)>(loc, builder);
+    break;
+#if LONG_DOUBLE == 80
```

Comment:
Please remove these `#ifdef` lines. The optimizer is target agnostic, portable, and may be used in a cross-compiler.

Reply:
In this case, it's better to forcibly declare both runtime entry points rather than pantomime what the header file is doing. The tradeoff is the linker emitting some unresolved symbols into the runtime vs. building a separate compiler for each target, which would violate some core LLVM design principles.

Reply:
@schweitzpgi, with the runtime reflection, I do not think it is possible to get rid of the ifdefs here...
It would lead to crashes when compiling f18 and looking for functions symbol not defined on the host... For instance, `RRSpacing16`  symbol is not defined in the runtime header of a X86 64 host (since it is ifdef out).
We'll have to find a solution for cross-compilation.

Reply:
Yes, I could not build bbc without the #ifdefs.

Reply:
Yes. That was what my comment "forcibly declare both runtime entry points" addresses. In the header file, these are under the same `#ifdef` conditional compilation. Here, in this runtime code, that should be subverted and both of them declared.